prog/dftb+/lib_dftb/encharges.F90

```diff
@@ -0,0 +1,695 @@
+
+    !> Neighbourlist.
+    integer, intent(in) :: iNeighbour(0:, :)
+
+    !> Square distances of the neighbours.
+    real(dp), intent(in) :: neighDist2(0:, :)
+
+    !> Mapping into the central cell.
+    integer, intent(in) :: img2CentCell(:)
+
+    !> Contains the points included in the reciprocal sum.
+    !  The set should not include the origin or inversion related points.
+    real(dp), allocatable, intent(in) :: recPoint(:, :)
+
+    !> Parameter for Ewald summation.
+    real(dp), intent(in) :: alpha
+
+    !> Volume of the real space unit cell.
+    real(dp), intent(in) :: volume
+
+    !> element-specific electronegativity
+    real(dp), intent(in) :: chi(:)
+
+    !> element-specific chemical hardnesses
+    real(dp), intent(in) :: gam(:)
+
+    !> element-specific CN scaling constant
+    real(dp), intent(in) :: kcn(:)
+
+    !> element-specific charge widths / atomic radii
+    real(dp), intent(in) :: rad(:)
+
+    !> Error function coordination number.
+    real(dp), intent(in) :: cn(:)
+
+    !> Derivative of the CN with respect to the Cartesian coordinates.
+    real(dp), intent(in) :: dcndr(:, :, :)
+
+    !> Derivative of the CN with respect to strain deformations.
+    real(dp), intent(in) :: dcndL(:, :, :)
+
+    !> Updated energy vector at return
+    real(dp), intent(inout), optional :: energies(:)
+
+    !> Updated gradient vector at return
+    real(dp), intent(inout), optional :: gradients(:, :)
+
+    !> Updated stress tensor at return
+    real(dp), intent(inout), optional :: stress(:, :)
+
+    !> Atomic partial charges.
+    real(dp), intent(out), optional :: qAtom(:)
+
+    !> Derivative of partial charges w.r.t. cartesian coordinates.
+    real(dp), intent(out), optional :: dqdr(:, :, :)
+
+    !> Derivative of partial charges w.r.t. strain deformations.
+    real(dp), intent(out), optional :: dqdL(:, :, :)
+
+    real(dp), parameter :: small = 1.0e-14_dp
+
+    !> Matrix of 1/R values for each atom pair.
+    real(dp), allocatable :: aMat(:, :)
+    real(dp), allocatable :: xVec(:)
+    real(dp), allocatable :: xFac(:)
+    real(dp), allocatable :: qVec(:)
+
+    real(dp), allocatable :: aInv(:, :)
+    integer, allocatable :: ipiv(:)
+
+    real(dp), allocatable :: aMatdr(:, :, :)
+    real(dp), allocatable :: aMatdL(:, :, :)
+    real(dp), allocatable :: aTrace(:, :)
+
+    logical :: tPeriodic
+    integer :: nDim
+    integer :: iAt1, iSp1, ii, jj
+    real(dp) :: tmp
+
+    tPeriodic = allocated(recPoint)
+
+    nDim = nAtom + 1
+    allocate(xVec(nDim), xFac(nAtom), aMatdr(3, nAtom, nDim), aMat(nDim, nDim), &
+        & aInv(nDim, nDim), qVec(nDim), aMatdL(3, 3, nDim), aTrace(3, nAtom))
+    allocate(ipiv(nDim))
+
+    ! Step 1: contruct RHS for linear system
+    do iAt1 = 1, nAtom
+      iSp1 = species(iAt1)
+      tmp = kcn(iSp1) / (sqrt(cn(iAt1)) + small)
+      xVec(iAt1) = -chi(iSp1) + tmp*cn(iAt1)
+      xFac(iAt1) = -0.5_dp*tmp
+    end do
+    xVec(nDim) = charge
+    xFac(nDim) = 0.0_dp
```

Comment:
Note: this writes beyond the boundaries of `xFac` (triggers run-time error with GNU compiler when compiled in debug mode).

Reply:
Removed.